### ISSUE TITLE
Lab 2 - EA set covering review

### ISSUE THREAD

**Issue by moDal7** [scoleri-mr/computational_intelligence_2022_301841#2]
# Lab 2 review
First of all, thanks a lot for the presentation in class, it was incredibly useful and insightful. I have read the code trying to review it and i found it very complete and exhaustive. I have some small remarks but they are mostly prompts for discussion, which I will write here. 

### Strategy remarks
- The _namedtuple_ is a very effective way of recording the genome and the values of fitness, very nice idea
- Wouldn't be a very effective way to kickstart the search to have a initial population composed only of valid solutions? 
- I am not sure if the _is_valid_ function is used anywhere, and if it is not, is the evaluate function verifying also the completeness of the individual as a solution after evaluating its fitness? I am not familiar with the Counter in Python, so there I might lose how it works
- I remember that you talked about the inefficiency of the fusion crossover despite its heaviness in computational terms. The fitness ratio is a factor that favors better individuals, making the mutation more "hill-climbing" like than random, especially in case of very fit individual vs. a very unfit one, in that case you would have a very small amount of mutations and basically make it a single element random mutation. Nevertheless I think you are the only ones to have tried to implement a method found in a scientific paper, and this is already much more than I did, so kudos 👍 
- The self-adapting mutation probability is greatly implemented.
- I think the plateau check is a bit "greedy" especially as a fixed factor: with higher N, it is likely to not have improvement for 30 straight generations. Maybe you could adapt it with respect to the problem original size, something like (N/20) + 20. Just an idea.
- The variable population size, for example, is a good use of a parameter that varies based on the problem original size.

### Formal remarks

- In the imports at the beginning _reduce from functools_ is not accessed and can be removed
- The is_valid function doesn't make use of the parameter P, which can be removed from the function declaration
- In the _initialize_population_ function I suppose the genome can be directly initialized as Tuple and have elements added to it
- The parameter n in the _evaluate_ function is not accessed and used
- The parameter _len_P_ in the crossover function isn't the same as _len(genome)_? Really a small thing but would make the function more self-explanatory with just the genomes as parameters

In conclusion, thanks a lot for the complete, clear and well-organized solution, with the graphs too! It was challenging to understand it all but congratulations on the huge work.

**Comment by scoleri-mr** [comment]
Thank you so much for your review! It's really helpful, complete and solid.    
I changed a few functions last minute and forgot to clean up some things like the *is_valid* which is not used. I did not implement another way to check if the solution is complete because the fitness function returns a tuple in which the first term is the number of elements discovered. I noticed that after a few generations that number reached N and never decreased again so I thought it was redundant to write some kind of check.  
The initialize_population function is written that way so that by simply changing the threshold value you can change the percentage of 1s that the individual in the initial population will have in their genome. But you are right, it definitely could be written more compactly.   
You are also right about the plateau check, I chose 30 because I plotted the same plot you saw at the end of my file for N=1000 and it still converged very early (under 80 generations) so I didn't feel like 30 was too restrictive. Still, it definitely needs to be considered as you suggested, especially if we were to increase N to larger values.  
Anyways thanks a lot for all your suggestions, I will definitely look into each of them and try to improve my code.
I really appreciate the work you did in understanding my code and finding several ways to improve it!
